proposals/nnnn-freestanding-macros.md

```diff
@@ -0,0 +1,322 @@
+
+```swift
+#line + 1
+#line as Int?
+```
+
+### Restrictions
+
+Like attached peer macros, a freestanding declaration macro can expand to any declaration that is syntatically and semantically well-formed within the context where the macro is expanded. It shares the same requirements and restrictions:
+
+- [**Specifying newly-introduced names**](https://github.com/apple/swift-evolution/blob/main/proposals/0389-attached-macros.md#specifying-newly-introduced-names)
+  - Note that only `named(...)` and `arbitrary` are allowed as macro-introduced names for a declaration macro. `overloaded`, `prefixed`, and `suffixed` do not make sense when there is no declaration from which to derive names.
+- [**Visibility of names used and introduced by macros**](https://github.com/apple/swift-evolution/blob/main/proposals/0389-attached-macros.md#visibility-of-names-used-and-introduced-by-macros)
+- [**Permitted declaration kinds**](https://github.com/apple/swift-evolution/blob/main/proposals/0389-attached-macros.md#permitted-declaration-kinds)
+
+One additional restriction is that a macro declaration can have at most one freestanding macro role. This is because top level and function scopes allow a combination of expressions, statements, and declarations, which would be ambiguous to a freestanding macro expansion with multiple roles.
+
+```swift
+@freestanding(expression)
+@freestanding(declaration) // error: a macro cannot have multiple freestanding macro roles 
+macro foo()
+```
+
+### Examples
+
+#### SE-0196 `warning` and `error`
+
+The `#warning` and `#error` directives introduced in [SE-0196](https://github.com/apple/swift-evolution/blob/main/proposals/0196-diagnostic-directives.md): can be implemented directly as declaration macros:
+
+```swift
+/// Emit a warning containing the given message.
+@freestanding(declaration) macro warning(_ message: String)
+
+/// Emit an error containing the given message
+@freestanding(declaration) macro error(_ message: String)
+```
+
+### Boilerplate generation
+
+Freestanding declaration macros can be used to generate boilerplace code. For example, the Standard Library could use such a macro to generate integer types.
+
+```swift
+@freestanding(declaration, names: arbitrary)
+fileprivate macro IntegerTypes(bitWidths: Int...)
+
+#IntegerTypes(bitWidths: 8, 16, 32, 64)
+```
+
+This expands to:
+
+```swift
+public struct Int8 { ... }
+public struct UInt8 { ... }
+
+public struct Int16 { ... }
+public struct UInt16 { ... }
+
+public struct Int32 { ... }
+public struct UInt32 { ... }
+
+public struct Int64 { ... }
+public struct UInt64 { ... }
+```
+
+### Template code generator
```

Comment:
Minor wording fix: The previous subsection uses "boilerplate generation" so I figured it's better to stick with "generation" rather than "generator".

```suggestion
### Template code generation
```